### ISSUE TITLE
Pre-release external-claim audit for v4.9.0 (tag gate)

### ISSUE THREAD

**Issue by dknauss** [dknauss/Sudo#421]
**The last gate before `v4.9.0` is tagged.** `AGENTS.md` → *Pre-release audit* requires every external claim added or modified since the last tag to be re-verified against live source, with findings appended to `docs/llm-lies-log.md`.

## Take this in a FRESH session, and do nothing else first

This is not ceremony. The audit is a long sequence of "open the upstream source and check", and the cheap failure at every step is to answer from what you already believe. That failure mode gets *worse* as context fills, so a session that has been running for hours is the worst possible one to run it — including the session that wrote this issue.

Tonight produced **seven** logged confabulations across four sessions (`llm-lies-log.md` #52–#58) with plenty of context headroom. Several were exactly this shape: a grep answering an adjacent question, a symbol name supplied from expectation, a file path asserted without opening it.

## The tree is frozen

Release candidate: **`39bcbd3`** on `main`. All §1 pre-tag gates verified green against it:

| gate | result |
|---|---|
| Version sync (5 points) | `4.9.0` |
| `blueprint.json` | `v4.8.0` — correct pre-tag |
| Unit | 1238 tests, 3680 assertions |
| PHPCS | 24/24 |
| PHPStan L6 | no errors |
| Psalm | 0 |
| `verify:metrics` | in sync |
| `verify:i18n` | in sync |
| `verify:sources` | 44 citations, snippets present |
| CHANGELOG / readme | both `4.9.0 - 2026-07-26` |

## Scope

```
git log v4.8.0..39bcbd3
```

Every external claim **added or modified** in that range. That means claims about WordPress core, other plugins, browser/platform behaviour, specifications, and third-party statistics.

**Verify against the named upstream source — not against `docs/upstream-sources.md`.** That registry is itself a set of claims; auditing a claim against a copy of the claim proves nothing.

Concentrate on files that carry behavioural prose: `docs/core-action-gate-proposal.md`, `docs/core-sudo-gate-implementation-spec.md`, `docs/security-model.md`, `docs/upstream-sources.md`, `CHANGELOG.md`, `readme.txt`, and comments in `includes/`.

## Known traps, all hit tonight

- **A snippet can confirm a positive fact but cannot falsify an absence** (#388). Rows claiming "no `apply_filters` to veto through" are unverifiable by the checker — re-grep them yourself.
- **`docs/` ships in the release zip** (see #419). Every claim in it is a claim in the distributed package, which is why a false core citation was a release-blocking fix (#420).
- **`docs/upstream-sources.md` is not on docs-lint's excluded list**, so an ISO date in a new row turns the workflow red. Phrase dates as "as read in July 2026".
- Registry IDs use the `GB-` prefix even for non-Gutenberg WordPress sources; widening that is tracked separately (#361).

## Done when

Findings appended to `docs/llm-lies-log.md` — appended, never rewritten — and either no unverified claims remain in range, or each surviving one is corrected or narrowed to something checkable.

Then the tag is maintainer-owned: cut annotated `v4.9.0`, let `release.yml` publish, and **only then** bump `blueprint.json` to the `v4.9.0` ZIP. A pre-tag bump points the public Playground demo at a ZIP that does not exist.

**Comment by dknauss** [comment]
**Release-candidate SHA corrected: `39bcbd3` → whatever is HEAD when the audit starts.**

That SHA is stale — `main` has since taken #395 (`09a14c7`). Rather than record another SHA that goes stale, the rule is:

> **Do not start this audit until the integrator declares the tree frozen and posts the final SHA here.** Audit that exact range, once.

Auditing a moving tree means either re-auditing or shipping a report about a tree that no longer exists.

The gate table in the issue body was measured against `39bcbd3` and should be re-run against the final SHA — it takes about two minutes:

```bash
composer test:unit && composer lint && composer analyse:phpstan && \
composer analyse:psalm && composer verify:metrics && composer verify:i18n && \
composer verify:sources
```

Everything else in the brief stands.

**Comment by dknauss** [comment]
**Known false-positive risk during the audit: the six `FT-*` rows fetch a moving branch.**

```
FT-SUDO-MODE, FT-SESSION-TIMEOUTS, FT-SESSION-DROPIN,
FT-MU-LOADER, FT-PROTECTED-CAPS, FT-EULA
  → https://raw.githubusercontent.com/snicco/fortress/beta/...
```

All six point at `beta`, not an immutable commit. If upstream pushes to that branch mid-audit, `verify:sources` goes red on content that has nothing to do with this release.

**If that happens: it is not a claim defect and not a release blocker.** `verify:sources` is not among `main`'s six required checks. Note it and move on; do not chase it, and do not let it delay the tag.

Pinning those rows to immutable commits is the correct fix and is **post-release** — it is part of a wider decision about Fortress's role, filed separately. Doing it now would mean resolving six commit SHAs and re-verifying during a freeze, which is exactly the late churn the freeze exists to prevent.

**Comment by dknauss** [comment]
## Findings from the #429 / [#433](https://github.com/dknauss/Sudo/pull/433) work, staged for the lies-log append

Posting these here rather than holding them in a session, per this issue's own instruction that a long-running session is the wrong place to run the audit. **I am that session** — this one has been going for hours — so I am handing the audit off rather than doing it, and leaving the raw material.

Numbering deliberately unassigned: the last append collided by assuming the next free number.

---

### A. A guard whose condition can never be true — five instances, three of them in the test written to catch the previous one

1. `stash_posted_to_its_own_screen()` compared the stashed `return_url` to the request URL. `wp_get_referer()` discards a referer equal to `REQUEST_URI`, and core's `user-new.php` forms carry no `action` attribute, so they self-post and it returns `false`. `return_url` is therefore `''` for the exact form #429 was about. The unit fixture supplied a value `Request_Stash::save()` cannot produce, so the test asserted correct behaviour about an impossible request.
2. The replacement gated the network `edit.php` clause on `is_network_admin()`. The landing runs under `admin-ajax.php`, which never calls `set_current_screen()` and does not define `WP_NETWORK_ADMIN` — false there even for a genuine network action. `tests/TestCase.php` also stubs it false suite-wide.
3. The guard test written to close the fail-open handler list read `$rule['pagenow']` / `$rule['method']`, but those live under `$rule['admin']`. It walked **zero rules** and asserted an empty array against an empty array — green while `user-edit.php` was still missing from the list it policed.
4. After that was fixed the walk still covered only the single-site half: `Action_Registry::rules()` merges network rules only when `is_multisite()` is true, which the suite stubs false. 13 of 15 `pagenow` values.
5. Resolving each `pagenow` to a full URL was still insufficient, because `$verified_screens` held bare filenames — so the `edit.php` entry absolved the network URL too, and disabling the network clause in production left the guard green.

**Rule:** a guard is not verified by reading it. If no mutation of the guarded code turns a test red, the test is not testing the guard. Every one of these was found by mutation; none by review of the code alone.

---

### B. A security invariant stated more broadly than it was proved

Claimed that using `return_url` only as a classifier meant "the set of reachable targets does not grow", and concluded the design was safe. The target set was indeed unchanged — but the requester gained *selection control* over which branch ran, and the branch they could select was the one the sibling GET path refuses for stated security reasons.

**Rule:** "X cannot grow" is not "the attacker gains nothing". Name what they gain, then argue it is acceptable.

---

### C. Prose describing behaviour the code did not have

The `is_handler_endpoint()` docblock and the CHANGELOG both said a guard test "closes that by walking Action_Registry's own pagenow set" — written in the same commit as the walk that read the wrong key and covered nothing, and left standing through a revision that still covered only half the registry.

**Rule:** a sentence about what a test covers is a claim with a falsifying command. Run it before writing the sentence.

---

### D. A release-baseline claim accepted from another session, then found wrong

Another session recorded in #429 as published fact that the Dashboard landing was "not new to 4.9.0 — pre-existing from #322 v1 (#350)". Two errors in one sentence: conflating *already on main* with *already released* (#350 is unreleased, so 4.9.0 is its first release), and asserting prior-release behaviour from context rather than running `git show`. `git show v4.8.0:includes/class-challenge.php` shows the redacted-POST branch redirecting to the originating form.

**Honest note:** I checked it only because I doubted the conclusion. Had I agreed, the regression ships. That is not process working.

**Rule this suggests, and which `AGENTS.md` does not currently contain:** a claim about *what a previous release did* is an external claim and needs the same `git show` discipline as an upstream citation.

---

### E. `--force-with-lease` reported success while discarding a commit

Recorded because it is the same failure family and cost a real commit tonight. The bare form compares against your **remote-tracking ref**, not the remote, so any `git fetch` between someone else's push and your force-push silently arms it. A session dropped `951bf42` (a merge of main) from `fix/322-rule-tester-accuracy` this way.

Independently verified as recoverable: no line unique to `951bf42` is absent from the current head. Nothing permanently lost.

**Rule:** pin the lease — `--force-with-lease=<ref>:<sha-you-reviewed>` — and after any force-push to a shared branch, diff the new head against the pre-push SHA in **both** directions before calling it clean.

---

### F. Reported by Security, and the most instructive of the set

A previous round fixed a scope-word over-claim in `developer-reference.md`, then grepped for the claim and fixed five copies. `CHANGELOG.md:7` **was in those grep results**. That line carried two defects — a wrong UI label and a `none of which` over-claim — and the label was fixed, the line seen, and the hit counted as discharged.

**Rule:** grepping harder would not have helped. A line is not discharged because you edited it; re-read what else it says.

---

## Note for whoever runs this audit

The frozen tree named above (`39bcbd3`) and its metrics table are **stale** — `main` has moved through #428, [#433](https://github.com/dknauss/Sudo/pull/433) and (pending) #383. I will post the final SHA and refreshed gate table once #383 merges. Do not audit against `39bcbd3`.

New external claims since that snapshot are the eight `GB-*` rows added by #433 (`GB-REFERER-SELFPOST`, `GB-USER-NEW-SELFPOST`, `GB-ADMIN-POST-BLANK`, `GB-OPTIONS-ALLSETTINGS`, `GB-ADMIN-PHP-BLANK`, `GB-NETWORK-EDIT-REDIRECT`, `GB-USER-EDIT-DIES`, `GB-UPDATE-NEEDS-ACTION`). Each was checked against `wordpress-develop` trunk when written and `composer verify:sources` passes — but per this issue's own rule, **verify against the named upstream source, not against the registry**, since the registry is itself the set of claims under audit.

**Comment by dknauss** [comment]
## Release candidate refreshed — audit against `b30691a`, not `39bcbd3`

`main` moved through four merges since the snapshot in the issue body. **The frozen tree named there is stale and its gate table is stale with it** (it reports 1238 tests; the tree now has 1248). Auditing against `39bcbd3` would check the wrong code.

New candidate: **`b30691a`**

```
b30691a fix(322): the Rule Tester must not predict replay at all (#383)
eb6fe3f fix(429): return a refused POST to its originating screen, not the dashboard (#433)
b50d0af docs(sources): GB-MS-SIGNUP-ROLE cited a line in the wrong function (#421) (#428)
09a14c7 docs: require a concurrent-work check before branching (#395)
```

### §1 pre-tag gates, re-run against `b30691a`

Run in a clean detached worktree with its own `composer install` — not a symlinked `vendor/`. That matters: during #433 a symlinked `vendor/` made the suite load `includes/` from the *other* worktree, so edits appeared to have no effect and a deliberately introduced parse error did not fail. Any gate run whose vendor is shared is not measuring the tree you think it is.

| gate | result |
|---|---|
| Version sync — `wp-sudo.php` header, `WP_SUDO_VERSION`, `tests/bootstrap.php`, `phpstan-bootstrap.php` | all `4.9.0` |
| `readme.txt` Stable tag | `4.9.0` |
| `blueprint.json` | `v4.8.0` — correct pre-tag; bump **after** the tag |
| Unit | **1248 tests, 3722 assertions** |
| PHPCS | 0 errors |
| PHPStan L6 + Psalm | 0 errors |
| `verify:metrics` | in sync |
| `verify:i18n` | in sync |
| `verify:sources` | **52 citations**, all snippets present |

### One item deliberately NOT changed

`CHANGELOG.md:3` and `readme.txt:198` both read `4.9.0 - 2026-07-26`. It is now **2026-07-27**, so the date is already wrong, and it will be wrong again if the tag slips another day.

I have left it alone on purpose: the correct value is the date the tag is actually cut, which is a maintainer action on a maintainer's schedule. Guessing it right today and wrong tomorrow is worse than flagging it. **Set both to the real tag date immediately before tagging.**

### Scope for the audit

`git log v4.8.0..b30691a` — every external claim added or modified in that range.

Since the stale snapshot, the new external claims are the eight `GB-*` rows added by [#433](https://github.com/dknauss/Sudo/pull/433):

`GB-REFERER-SELFPOST` · `GB-USER-NEW-SELFPOST` · `GB-ADMIN-POST-BLANK` · `GB-OPTIONS-ALLSETTINGS` · `GB-ADMIN-PHP-BLANK` · `GB-NETWORK-EDIT-REDIRECT` · `GB-USER-EDIT-DIES` · `GB-UPDATE-NEEDS-ACTION`

Each was checked against `wordpress-develop` trunk when written, and `verify:sources` passes. **Neither fact discharges the audit** — per this issue's own rule, verify against the named upstream source, because the registry is itself the set of claims under audit and `verify:sources` only proves a snippet still exists at a URL, not that it supports the claim attached to it.

Also worth re-reading rather than trusting: `#383` corrected a `none of which` over-claim in `CHANGELOG.md:7` that had survived a previous sweep **because the line appeared in that sweep's grep results and was counted as handled after a different defect on the same line was fixed.** Two defects, one line, one fix. Assume other lines edited during this release carry the same pattern.

The findings from tonight's work are in the comment above, staged for the `llm-lies-log.md` append that closes this issue.
